azure/datalake/store/transfer.py

```diff
@@ -240,9 +245,14 @@ def __init__(self, adlfs, transfer, merge=None, nthreads=None,
         self._chunked = chunked
         self._unique_temporary = unique_temporary
         self._unique_str = uuid.uuid4().hex
+        self._progress_callback=progress_callback
+        self._progress_lock = threading.Lock()
```

Comment:
Is there an existing benchmark for this that I could test with?

IMHO the lock contention penalty should be small compared to the IO generated by transfer which will be a non-trivial HTTP operation so I'd expect this to have a small impact if any.

That said, `_update` is rather elaborated and has multiple operations updating shared state. On a first look it doesn't seem necessarily problematic but something like this:

```
 self._chunks[obj]['actual'] = nbytes
 self._chunks[obj]['exception'] = exception
```
won't execute atomically. 

Ultimately, if this locking doesn't affect performance, perhaps may be more appropriate to happen in `_update` rather than `_update_progress`.